File: src/parsers/aspnet-parser.ts
import * as cheerio from 'cheerio';
import type { AspNetWebFormData } from '../types/session';
import { ParsingError } from '../core/errors';
import { ASPNET_SELECTORS, ROOT_FORM_SELECTORS } from '../constants/selectors';

export function extractAspNetFormData(html: string): AspNetWebFormData {
  const $ = cheerio.load(html);

  const viewstate = $(ASPNET_SELECTORS.VIEWSTATE).val();
  const eventValidation = $(ASPNET_SELECTORS.EVENTVALIDATION).val();
  const viewstateGenerator = $(ASPNET_SELECTORS.VIEWSTATEGENERATOR).val();
  const previousPage = $(ASPNET_SELECTORS.PREVIOUSPAGE).val();

  if (typeof viewstate === 'string' && viewstate) {
    return {
      __VIEWSTATE: viewstate,
      __EVENTVALIDATION: typeof eventValidation === 'string' ? eventValidation : '',
      __VIEWSTATEGENERATOR: typeof viewstateGenerator === 'string' ? viewstateGenerator : undefined,
      __PREVIOUSPAGE: typeof previousPage === 'string' ? previousPage : undefined,
    };
  }

  const rootViewstate = $(ROOT_FORM_SELECTORS.VIEWSTATE).val();
  const rootEventValidation = $(ROOT_FORM_SELECTORS.EVENTVALIDATION).val();
  const rootViewstateGenerator = $(ROOT_FORM_SELECTORS.VIEWSTATEGENERATOR).val();

  if (typeof rootViewstate === 'string' && rootViewstate) {
    return {
      __VIEWSTATE: rootViewstate,
      __EVENTVALIDATION: typeof rootEventValidation === 'string' ? rootEventValidation : '',
      __VIEWSTATEGENERATOR: typeof rootViewstateGenerator === 'string' ? rootViewstateGenerator : undefined,
    };
  }

  throw new ParsingError(
    'Failed to extract form data: no __VIEWSTATE or __EVENTVALIDATION found. Page may be a redirect/captcha.',
  );
}

export function extractRootFormData(html: string): {
  __VIEWSTATE: string;
  __EVENTVALIDATION: string;
  __VIEWSTATEGENERATOR?: string;
  studentId: string;
  password: string;
  recaptchaResponse: string;
  recaptchaSiteKey: string;
  action: string;
} {
  const $ = cheerio.load(html);

  const viewstate = $(ROOT_FORM_SELECTORS.VIEWSTATE).val();
  const eventValidation = $(ROOT_FORM_SELECTORS.EVENTVALIDATION).val();
  const viewstateGenerator = $(ROOT_FORM_SELECTORS.VIEWSTATEGENERATOR).val();

  if (typeof viewstate !== 'string' || typeof eventValidation !== 'string') {
    throw new ParsingError('Failed to extract root form data: missing hidden fields');
  }

  const formAction = $('form').attr('action') || './';
  const recaptchaSiteKey = html.match(/recaptcha\/api\.js\?render=([A-Za-z0-9_-]+)/)?.[1] || '';

  return {
    __VIEWSTATE: viewstate,
    __EVENTVALIDATION: eventValidation,
    __VIEWSTATEGENERATOR: typeof viewstateGenerator === 'string' ? viewstateGenerator : undefined,
    studentId: 'txtStudentID',
    password: 'txtPassword',
    recaptchaResponse: 'g-recaptcha-response',
    recaptchaSiteKey,
    action: formAction,
  };
}

export function buildLoginData(
  formData: { __VIEWSTATE: string; __EVENTVALIDATION: string; __VIEWSTATEGENERATOR?: string },
  username: string,
  password: string,
): Record<string, string> {
  return {
    __VIEWSTATE: formData.__VIEWSTATE,
    __EVENTVALIDATION: formData.__EVENTVALIDATION,
    ...(formData.__VIEWSTATEGENERATOR
      ? { __VIEWSTATEGENERATOR: formData.__VIEWSTATEGENERATOR }
      : {}),
    txtStudentID: username,
    txtPassword: password,
    ibtnLogin: 'Sign In',
  };
}

export function buildPostbackData(
  formData: AspNetWebFormData,
  extraFields?: Record<string, string>,
): Record<string, string> {
  const data: Record<string, string> = {
    __VIEWSTATE: formData.__VIEWSTATE,
    __EVENTVALIDATION: formData.__EVENTVALIDATION,
  };

  if (formData.__VIEWSTATEGENERATOR) {
    data.__VIEWSTATEGENERATOR = formData.__VIEWSTATEGENERATOR;
  }

  if (formData.__PREVIOUSPAGE) {
    data.__PREVIOUSPAGE = formData.__PREVIOUSPAGE;
  }

  if (extraFields) {
    for (const [key, value] of Object.entries(extraFields)) {
      data[key] = value;
    }
  }

  return data;
}

export function isLoginSuccess(html: string): boolean {
  return (
    !html.includes('Incorrect') &&
    !html.includes('Invalid') &&
    !html.includes('Login.aspx') &&
    html.includes('__VIEWSTATE')
  );
}

export function isLoginError(html: string): boolean {
  return (
    html.includes('Incorrect') ||
    html.includes('Invalid') ||
    html.includes('error') ||
    html.includes('alert-danger')
  );
}

File: vulms_dart/lib/src/parsers/aspnet_parser.dart
import 'package:html/dom.dart';
import 'package:html/parser.dart' as html_parser;

import '../models/session.dart';
import '../exceptions/exceptions.dart';
import '../utils/constants.dart';

/// Parses ASP.NET form data from HTML pages.
///
/// Extracts __VIEWSTATE, __EVENTVALIDATION, and other hidden fields
/// required for ASP.NET PostBack requests.
class AspNetParser {
  AspNetParser._();

  /// Extract ASP.NET form data from HTML.
  ///
  /// Throws [ParsingException] if __VIEWSTATE cannot be found.
  static AspNetFormData extractFormData(String html) {
    final doc = html_parser.parse(html);

    // Try standard selectors first
    var viewState = _extractValue(doc, AspNetSelectors.viewState);
    var eventValidation = _extractValue(doc, AspNetSelectors.eventValidation);
    var viewStateGenerator =
        _extractValue(doc, AspNetSelectors.viewStateGenerator);
    final previousPage = _extractValue(doc, AspNetSelectors.previousPage);

    // Fallback to root form selectors
    viewState ??= _extractValue(doc, AspNetSelectors.rootViewState);
    eventValidation ??= _extractValue(doc, AspNetSelectors.rootEventValidation);
    viewStateGenerator ??=
        _extractValue(doc, AspNetSelectors.rootViewStateGenerator);

    if (viewState == null || viewState.isEmpty) {
      throw const ParsingException(
        'Failed to extract form data: no __VIEWSTATE found. '
        'Page may be a redirect or captcha.',
      );
    }

    return AspNetFormData(
      viewState: viewState,
      eventValidation: eventValidation ?? '',
      viewStateGenerator: viewStateGenerator,
      previousPage: previousPage,
    );
  }

  /// Build login form data from extracted form fields.
  static Map<String, String> buildLoginData(
    AspNetFormData formData,
    String username,
    String password,
  ) {
    final data = <String, String>{
      '__VIEWSTATE': formData.viewState,
      '__EVENTVALIDATION': formData.eventValidation,
      'txtStudentID': username,
      'txtPassword': password,
      'ibtnLogin': 'Sign In',
    };
    if (formData.viewStateGenerator != null) {
      data['__VIEWSTATEGENERATOR'] = formData.viewStateGenerator!;
    }
    return data;
  }

  /// Build PostBack form data from extracted form fields.
  static Map<String, String> buildPostbackData(
    AspNetFormData formData, {
    Map<String, String>? extraFields,
  }) {
    final data = <String, String>{
      '__VIEWSTATE': formData.viewState,
      '__EVENTVALIDATION': formData.eventValidation,
    };
    if (formData.viewStateGenerator != null) {
      data['__VIEWSTATEGENERATOR'] = formData.viewStateGenerator!;
    }
    if (formData.previousPage != null) {
      data['__PREVIOUSPAGE'] = formData.previousPage!;
    }
    if (extraFields != null) {
      data.addAll(extraFields);
    }
    return data;
  }

  /// Check if the HTML indicates a successful login.
  static bool isLoginSuccess(String html) {
    return !html.contains('Incorrect') &&
        !html.contains('Invalid') &&
        !html.contains('Login.aspx') &&
        html.contains('__VIEWSTATE');
  }

  /// Check if the HTML indicates a login error.
  static bool isLoginError(String html) {
    return html.contains('Incorrect') ||
        html.contains('Invalid') ||
        html.contains('alert-danger');
  }

  static String? _extractValue(Document doc, String selector) {
    final element = doc.querySelector(selector);
    if (element == null) return null;
    return element.attributes['value'];
  }
}
